File: CodeWars/6kyu/JavaScript/016-array-dot-diff.js
// https://www.codewars.com/kata/523f5d21c841566fde000009
// Your goal in this kata is to implement a difference function, which subtracts one list from another and returns the result.
// It should remove all values from list a, which are present in list b.
// arrayDiff([1,2,2,2,3],[2]) == [1,3]

function arrayDiff(a, b) {
    //create a set of b array
    let filteringValues = new Set(b);

    return a.filter(value => !filteringValues.has(value))
}

console.log(arrayDiff([], [4, 5]), [], "a was [], b was [4,5]");
console.log(arrayDiff([3, 4], [3]), [4], "a was [3,4], b was [3]");
console.log(arrayDiff([1, 8, 2], []), [1, 8, 2], "a was [1,8,2], b was []");

File: CodeWars/6kyu/Python/016-array-dot-diff.py
# https://www.codewars.com/kata/523f5d21c841566fde000009
# Your goal in this kata is to implement a difference function, which subtracts one list from another and returns the result.
# It should remove all values from list a, which are present in list b keeping their order.
# array_diff([1,2],[1]) == [2]

def array_diff(a, b):
    result = []

    for value in a:
        if value not in b:
            result.append(value)

    return result


def array_diff(a, b):
    return [x for x in a if x not in b]


print(array_diff([1, 2], [1]), [2], "a was [1,2], b was [1], expected [2]")
print(array_diff([1, 2, 2], [1]), [2, 2],
      "a was [1,2,2], b was [1], expected [2,2]")
print(array_diff([1, 2, 2], [2]), [1],
      "a was [1,2,2], b was [2], expected [1]")
print(array_diff([1, 2, 2], []), [1, 2, 2],
      "a was [1,2,2], b was [], expected [1,2,2]")
print(array_diff([], [1, 2]), [], "a was [], b was [1,2], expected []")
print(array_diff([1, 2, 3], [1, 2]), [3],
      "a was [1,2,3], b was [1, 2], expected [3]")
